src/plugins/tangorin.rs

```diff
@@ -0,0 +1,214 @@
+use std::io;
+use irc::client::prelude::*;
+use regex::Regex;
+use plugin::Plugin;
+
+extern crate kuchiki;
+use kuchiki::traits::*;
+
+register_plugin!(Tangorin);
+
+lazy_static! {
+    static ref RE: Regex = Regex::new(r"!tangorin (\S+)").unwrap();
+}
+
+impl Tangorin {
+    fn grep_kanji(&self, msg: &str) -> Option<String> {
+        match RE.captures(msg) {
+            Some(captures) => {
+                Some(format!("{}", captures.at(1).unwrap()))
+            },
+            None => None
+        }
+    }
+
+    fn retrieve_from_selector(&self, doc: &kuchiki::NodeRef, selector: &str) -> Option<String> {
+        if let Some(match_) = doc.select(selector).unwrap().next() {
+            let node    = match_.as_node().first_child().unwrap();
+            let borrowed_text   = node.as_text().unwrap().borrow();
+            let mut retrieved_text = borrowed_text.clone();
+            retrieved_text = retrieved_text.trim().to_string();
+            
+            Some(retrieved_text)
+        }
+        else{
+            None
+        }
+    }
+
+    fn retrieve_meaning(&self, doc: &kuchiki::NodeRef) -> Option<String> {
+        if let Some(match_) = doc.select("span[class=eng]").unwrap().next() {
+            let node = match_.as_node(); 
+            let children = node.children();
+            let mut meaning = "".to_string();
+            for child_ in children {
+                if let Some(text_) = child_.as_text() {
+                    meaning.push_str(text_.borrow().as_str());
+                }
+                else{
+                    if let Some(text_) = self.inner_text(&child_) {
+                        meaning.push_str(text_.as_str());
+                    }
+                }
+            }
+
+           Some(meaning)
+        }
+        else{
+            None
+        }
+    }
+
+    fn retrieve_info(&self, doc: &kuchiki::NodeRef) -> Option<String> {
+        if let Some(match_) = doc.select("span[class=eng]").unwrap().next() {
+            let node = match_.as_node();
+            let following_siblings = node.following_siblings();
+
+            if let Ok(mut info_sibs) = following_siblings.select("i[class=d-info]") {
+                match info_sibs.next() {
+                    Some(first_sibling) => self.inner_text(first_sibling.as_node()),
+                    None => None
+                }
+            }
+            else{
+                None
+            }
+        }
+        else {
+            None
+        }
+    }
+
+    fn inner_text(&self, root: &kuchiki::NodeRef) -> Option<String> {
+        if let Some(match_) = root.first_child() {
+            let res = match match_.as_text() {
+                Some(result_str) => result_str.borrow(),
+                None => { return None }
+            };
+            Some(res.clone())
```

Comment:
prefer `to_owned()` in place of `clone()`. Btw I think this block can be make shorter